File: dashboard/src/api/statistics.ts
import { useQuery } from '@tanstack/react-query';

import { httpGet } from './httpClient';

export type DashboardSummary = {
  vehicles_in_parking: number;
  today_checkins: number;
  today_checkouts: number;
  today_revenue: number;
  monthly_revenue: number;
  pending_invoices: number;
  active_subscription_users: number;
  expiring_subscriptions: number;

  // backward-compatible (optional)
  total_users?: number;
  registered_users?: number;
  unregistered_users?: number;
  registration_rate?: number;
  active_subscriptions?: number;
  today_sessions?: number;
};

export type ChartPoint = {
  label: string;
  value: number;
};

export type MonthlySubscriptionPoint = {
  month: string;
  new_subscriptions: number;
};

export type MonthlyRevenuePoint = {
  month: string;
  revenue: number;
};

export type MonthlyParkingSessionPoint = {
  month: string;
  sessions: number;
};

export type StatusDistributionPoint = {
  status: string;
  count: number;
};

export type DashboardCharts = {
  subscription_adoption: ChartPoint[];
  monthly_subscriptions: MonthlySubscriptionPoint[];
  monthly_revenue: MonthlyRevenuePoint[];
  monthly_parking_sessions: MonthlyParkingSessionPoint[];
  invoice_status_distribution: StatusDistributionPoint[];
  subscription_status_distribution: StatusDistributionPoint[];
};

export type RecentInvoiceItem = {
  id: string;
  user_code?: string | null;
  user_full_name?: string | null;
  amount: number;
  status: string;
  payment_method?: string | null;
  created_at: string;
};

export type RecentParkingSessionItem = {
  id: string;
  user_code?: string | null;
  user_full_name?: string | null;
  license_plate?: string | null;
  status: string;
  check_in_time?: string | null;
  check_out_time?: string | null;
};

export type DashboardRecent = {
  invoices: RecentInvoiceItem[];
  sessions: RecentParkingSessionItem[];
};

export const useDashboardSummary = () => {
  return useQuery({
    queryKey: ['admin', 'dashboard', 'summary'],
    queryFn: () => httpGet<DashboardSummary>('/statistics/summary'),
    staleTime: 1000 * 30,
  });
};

export const useDashboardCharts = () => {
  return useQuery({
    queryKey: ['admin', 'dashboard', 'charts'],
    queryFn: () => httpGet<DashboardCharts>('/statistics/charts'),
    staleTime: 1000 * 60,
  });
};

export const useDashboardRecent = () => {
  return useQuery({
    queryKey: ['admin', 'dashboard', 'recent'],
    queryFn: () => httpGet<DashboardRecent>('/statistics/recent'),
    staleTime: 1000 * 20,
  });
};

File: backend/app/models/statistics.py
from __future__ import annotations

from datetime import datetime
from sqlmodel import SQLModel


class DashboardSummary(SQLModel):
    # Preferred KPIs for school parking operations
    vehicles_in_parking: int
    today_checkins: int
    today_checkouts: int
    today_revenue: int | float
    monthly_revenue: int | float
    pending_invoices: int
    active_subscription_users: int
    expiring_subscriptions: int

    # Backward-compatible fields (dashboard v1)
    total_users: int = 0
    registered_users: int = 0
    unregistered_users: int = 0
    registration_rate: float = 0
    active_subscriptions: int = 0
    today_sessions: int = 0


class ConsoleSummary(SQLModel):
    vehicles_in_parking: int
    today_checkins: int
    today_checkouts: int


class ChartPoint(SQLModel):
    label: str
    value: int | float


class MonthlySubscriptionPoint(SQLModel):
    month: str
    new_subscriptions: int


class MonthlyRevenuePoint(SQLModel):
    month: str
    revenue: int | float


class MonthlyParkingSessionPoint(SQLModel):
    month: str
    sessions: int


class StatusDistributionPoint(SQLModel):
    status: str
    count: int


class DashboardCharts(SQLModel):
    subscription_adoption: list[ChartPoint]
    monthly_subscriptions: list[MonthlySubscriptionPoint]
    monthly_revenue: list[MonthlyRevenuePoint]
    monthly_parking_sessions: list[MonthlyParkingSessionPoint]
    invoice_status_distribution: list[StatusDistributionPoint]
    subscription_status_distribution: list[StatusDistributionPoint]


class RecentInvoiceItem(SQLModel):
    id: str
    user_code: str | None = None
    user_full_name: str | None = None
    amount: int | float
    status: str
    payment_method: str | None = None
    created_at: datetime


class RecentParkingSessionItem(SQLModel):
    id: str
    user_code: str | None = None
    user_full_name: str | None = None
    license_plate: str | None = None
    status: str
    check_in_time: datetime | None = None
    check_out_time: datetime | None = None


class DashboardRecent(SQLModel):
    invoices: list[RecentInvoiceItem]
    sessions: list[RecentParkingSessionItem]
